ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryBuilder.java

```diff
@@ -650,4 +671,131 @@ private static void updateListProperty(
     valueList.add(value);
     properties.put(key, valueList);
   }
+
+  private static class TransientQuerySinkProcessor implements
+      Processor<Object, GenericRow, Void, Void> {
+
+    private final TransientQueryQueue queue;
+    private final Optional<ImmutableMap<TopicPartition, Long>> endOffsets;
+    private final ConcurrentHashSet<TopicPartition> donePartitions;
+    private ProcessorContext<Void, Void> context;
+
+    TransientQuerySinkProcessor(
+        final TransientQueryQueue queue,
+        final Optional<ImmutableMap<TopicPartition, Long>> endOffsets,
+        final ConcurrentHashSet<TopicPartition> donePartitions) {
+
+      this.queue = queue;
+      this.endOffsets = endOffsets;
+      this.donePartitions = donePartitions;
+    }
+
+    @Override
+    public void init(final ProcessorContext<Void, Void> context) {
+      Processor.super.init(context);
+      this.context = context;
+      if (endOffsets.isPresent()) {
+        // registering a punctuation here is the easiest way to check
+        // if we're complete even if we don't see all records.
+        // The in-line check in process() is faster,
+        // but some records may be filtered out before reaching us.
+        context.schedule(
+            Duration.ofMillis(100),
+            PunctuationType.WALL_CLOCK_TIME,
+            timestamp -> checkCompletion()
+        );
+      }
+    }
+
+    @Override
+    public void process(final Record<Object, GenericRow> record) {
+      final Optional<TopicPartition> topicPartition =
+          context
+              .recordMetadata()
+              .map(m -> new TopicPartition(m.topic(), m.partition()));
+
+      final boolean alreadyDone =
+          topicPartition.isPresent() && donePartitions.contains(topicPartition.get());
+
+      // We ignore null values in streams as invalid data and drop the record.
+      // If we're already done with the current partition, we just drop the record.
+      if (record.value() != null && !alreadyDone) {
```

Comment:
Do push queries not handle null data values?  For some reason I thought they did.

Reply:
Yeah, this code was here before, so I guess not.

Reply:
That is consistent with KS semantics, that null-values are considered invalid and always filtered out. In tables they are considered as tombstones of course, and we sort of letting the users be responsible of the difference between the two when converting in between.